src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,188 @@
+      while (i < nProducts) {
+        val leftMat = leftBlock(row, i, context)
+        val rightMat = rightBlock(i, col, context)
+        (leftMat, rightMat) match {
+          case (x: DenseMatrix, y: DenseMatrix) =>
+            multiplyAccumulateWithExtension(x, y, finalResult)
+          case _ =>
+             throw new SparkException(s"No support for multiplying: ${leftMat.getClass} by ${rightMat.getClass}.")
+        }
+
+        i += 1
+      }
+
+      Iterator.single(((row, col), finalResult))
+    }
+
+    protected def getPartitions: Array[Partition] =
+      partitions
+
+    /** Optionally overridden by subclasses to specify how they are partitioned. */
+    @transient override val partitioner: Option[Partitioner] =
+      Some(new GridPartitioner(rowBlocks, colBlocks, 1, 1))
+  }
+
+  case class IntPartition(index: Int) extends Partition { }
+
+  private def gridPartition(m: BlockMatrix): (BlockMatrix, GridPartitioner) = {
+    val p = GridPartitioner(m.numRowBlocks, m.numColBlocks)
+    (new BlockMatrix(m.blocks.partitionBy(p), m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols()),
+      p)
+  }
+
+  private def ensureGridPartitioning(m: BlockMatrix): (BlockMatrix, GridPartitioner) = l.blocks.partitioner match {
+    case Some(gp: GridPartitioner) if (gp.rowsPerPart == 1 && gp.colsPerPart == 1) =>
+      (l, gp)
+    case Some(gp: GridPartitioner) =>
+      logDebug(s"Repartitioning a matrix (slow), $l, with a grid partitioner that didn't have 1 block per partition, had: ${gp.rowsPerPart} x ${gp.colsPerPart}")
+      gridPartition(l)
+    case Some(p) =>
+      logDebug(s"Repartitioning a matrix (slow), $l, with a non-grid partitioner: $p")
+      gridPartition(l)
+    case None =>
+      logDebug(s"Partitioning a matrix (slow), $l, with a no partitioner")
+      gridPartition(l)
+  }
+
+  def multiply(preL: BlockMatrix, preR: BlockMatrix): BlockMatrix = {
+    require(preL.numCols() == preR.numRows(),
+      s"""The number of columns of theleft matrix and the number of rows of the right
+          matrix must be equal: ${l.numRows()} x ${l.numCols()}, ${r.numRows()}
+          x ${r.numCols()}. If you think they should be equal, try setting the
+          dimensions of A and B explicitly while initializing them.""".stripMargin)
+    // FIXME: don't require same blocksize on each matrix
+    require(preL.colsPerBlock == preR.rowsPerBlock,
+      s"""The number of columns in blocks of the left matrix and the number of rows in
+          blocks of the right matrix must be equal: ${l.rowsPerBlock} x
+          ${l.colsPerBlock}, ${r.rowsPerBlock} x ${r.colsPerBlock}. Generally,
+          all matrices should use square blocks of the same dimension.""".stripMargin)
+
+    val (l, lPartitioner) = ensureGridPartitioning(preL)
+    val (r, rPartitioner) = ensureGridPartitioning(preR)
+    val rowBlocks = l.numRowBlocks
+    val nProducts = l.numColBlocks
+    val colBlocks = r.numColBlocks
+    val deps = Array[Dependency[_]](
```

Comment:
override RDD.getDependencies inside BlockMatrixMultiplyRDD instead of computing this on the outside and passing it in.  I think that'll be cleaner.